### ISSUE TITLE
[PKI meta] Cert path building and verification needs many further improvements

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#13780]
Certificate path/chain _construction_ (see [RFC 4158](https://tools.ietf.org/html/rfc4158)) and _verification_ (see [RFC 5820 section 6](https://tools.ietf.org/html/rfc5280#section-6)) form the **critical core** of using public-key certificates. Their implementation is **inherently complex** due to a large variety of factors that play a role and strongly interact.

Obviously their implementations must be **correct** such that no illegal chain is accepted.
They should be **complete** (as far as reasonably possible) such that all legal chains are found and accepted.
Due to their criticality and complexity it is important also for maintainability that they are **cleanly coded**.
They should also be **well documented**, at least at high level, such that user understand 
what they can expect and which parameters and which options have which effect.

Regarding each of these (sub-)topics, here is my current view on them regarding the OpenSSL implementation:

* It appears pretty correct in the sense that despite a large body of test cases no insecure/illegal chain is known to be accepted.
While working myself through the code I recently found just one potential issue here: I think it would be good if chain verification re-checked the trust placed in the root of the chain, such that the security of chain verification becomes is independent of the checks done during chain building. This is requested to be added in #13770. 

* It is fairly complete, while there is a at least one long known bug: 
depending on the order of certs in the trust store, it can happen that an acceptable chain is not found (and thus verification is wrongly rejected) just because there is a (root) cert in the trust store that appears suitable because its subject matches but its public key does not, and the chain building selects this cert but then fails and does not backtrack its first choice, although there there is a further cert that would have been found later in the trust store with which an acceptable chain could be formed.
Since this issue came up again on the openssl-users mailing list recently, I invested quite some part of my spare time not only to analyze it but also to provide a fix in #13748. I believe that that PR includes pretty general improvements of the chain building algorithm and actually fixes a design flaw of `build_chain()`: it makes little sense to try finding an issuer of the current end of the chain if this end can be found immediately in the trust store (and is actually trusted).
That PR has already been receiving pretty strong pushback by more than one person although - or likely also because - I doubt that anyone has taken the time yet to look into it and understand its essence. Instead, there is the - understandable - general fear that doing any non-trivial change to the pretty complex and rather messy code can by mistake cause a regression anywhere in the overall implementation, which touches both the correctness and the cleanliness topic.
BTW, related to completeness are compatibility policies regarding implicit trust in certs that are members of the trust store. At least the "partial chains" option has been causing for quite a while misunderstandings and grief at the users' end and some internal discussions, in particular in #7871. 

* The code is obviously not as clean and readable as it should be. 
Despite some efforts several years back to improve on this, the code, in particular the chain building function including some auxiliary ones, shows (again?) symptoms of rank growth and is still extremely - and in party needlessly - hard to understand.
A further - partly unavoidable - complication is the DANE support. Although it is a special (TLS) case that is not needed in most certificate use cases, is implementation strongly influences the structure of the chain building and checking functions in a way that seems questionable at least to me.
Moreover, the code still supports obsolete options such as  `-trusted_first` and `no_alt_chains`, which needlessly complicates matters.
The code does not even strictly follow superficial coding guidelines. I've tackled this, among others, in the refactoring PR #13070 (which has not received much attention/support yet). Still that PR only touches the surface and is clearly not sufficient: 
More fundamental cleanup of the code structure is (over-)due. In particular, I suggest splitting (for both clarity and maintainability) the lengthy file `x509_vfy.c` into `x509_build_chain.c` and `x509_verify_chain.c`, refactoring the lengthy function `build_chain()`, changing some misleading names, and moving some checks to where they belong or are at least better to understand.

* Its current documentation is pretty spotty (and in few partly outdated), such that the global picture it provides to non-expert users is surprisingly vague, if not misleading. No surprise that most users are not even aware that simply placing certificates in the trust store is in general not sufficient for making them trusted and falls short of what the project recommends.
My PR #13735 aims at fixing at least the most pressing issues there. Though it already took quite some effort there are still several aspects that are insufficiently or not at all documented. In particular, there are undocumented flags and options. For instance, `X509_TRUST_NO_SS_COMPAT` and `X509_TRUST_DO_SS_COMPAT` may be just internally needed, but their role is obscure, in particular their relation with `X509_V_FLAG_PARTIAL_CHAIN` and with the (default) trust identifiers.
It looks like there are subtle differences between trust identifiers such as `X509_TRUST_SSL_CLIENT` used for trust anchors, purpose settings such as `sslclient` used for verifying certificates, and the externally defined X.509 extended key usages. Maybe different nomenclature such as "use", "purpose", and "trust" is used just by mistake and should be cleaned up, or is used for good reason but then needs to be properly distinguished and explained. Tautological documentation such as "X509_VERIFY_PARAM_set_trust() sets the trust setting in param to trust" is of little help since it remains unclear what its effect is, and what the difference to related functions and concepts such as X509_VERIFY_PARAM_set_purpose() is.
There is likely more such mistiness due to further aspects not (sufficiently) documented that I'm not aware of.

**Comment by richsalz** [comment]
> It appears pretty correct in the sense that despite a large body of test cases no insecure/illegal chain is known to be accepted.

I disagree about large body of test cases.  I skimmed over all test files that have `X509_STORE` and focusing on `verify_extra_test` it looks like there's under a dozen.

**Comment by beldmit** [comment]
I'd also add that the code does not provide enough information about a particular chain building. I think it's useful to have a way to get a step-by-step information about the chain of trust building process in some debugging mode but I'm afraid it's out of scope of the 3.0 milestone.

**Comment by DDvO** [comment]
> > It appears pretty correct in the sense that despite a large body of test cases no insecure/illegal chain is known to be accepted.
> 
> I disagree about large body of test cases. I skimmed over all test files that have `X509_STORE` and focusing on `verify_extra_test` it looks like there's under a dozen.

There is `25-test_verify.t` directly addressing the topic with some 150+ test cases.
There are many more tests that (at least indirectly) use cert verification and thus also chain building, 
such as SSL tests, CMP tests, OCSP tests, and the tests for various apps where verification also plays a role.
`fgrep verify test/*.c|wc` yields 220 hits,
`fgrep verify test/recipes/*.t|wc` yields 345.

**Comment by richsalz** [comment]
Thank you for the more detailed numbers.  Negative tests?  Mix of intermediate and end-entity certs in trust stores?

**Comment by DDvO** [comment]
All sorts of tests, surely most of them positive ones; I did not check in detail. 

`25-test_verify.t` contains many dedicated negative tests, of course. 
`fgrep 'ok(!' test/recipes/25-test_verify.t|wc` yields 87.

**Comment by DDvO** [comment]
> I'd also add that the code does not provide enough information about a particular chain building. I think it's useful to have a way to get a step-by-step information about the chain of trust building process in some debugging mode but I'm afraid it's out of scope of the 3.0 milestone.

Good point that a debug/log feature could be helpful. It should not be hard to add.
For instance, `crypto/cmp/cmp_vfy.c` has pretty neat diagnostic features,
which have proved very helpful for finding bugs during development and application configuration,
which can be challenging for novices, in particular when multiple CA hierarchies are involved.

**Comment by richsalz** [comment]
So I guess there are many tests, but we don't know specifics.  Thanks.

**Comment by DDvO** [comment]
The tests in `25-test_verify.t`are pretty readable, so everyone can check.
At some earlier time I convinced myself that they do make sense.
I have not investigated in their systematic completeness, though.

**Comment by richsalz** [comment]
> The tests in `25-test_verify.t`are pretty readable, so everyone can check.
> At some earlier time I convinced myself that they do make sense.
> I have not investigated in their systematic completeness, though.

The tests are readable, but you can't know *what* they test without inspecting the certs involved, which is tedious at best and time-consuming.

**Comment by vdukhovni** [comment]
> * While working myself through the code I recently found just one potential issue here: I think it would be good if chain verification re-checked the trust placed in the root of the chain, such that the security of chain verification becomes is independent of the checks done during chain building. This is requested to be added in #13770.

This is a mistake, resulting from a misunderstanding of the role of the `internal_verify()` function.  Its purpose is solely to check chain *integrity*.  That is, to ensure that the chain is actually a chain in which each issuer has applied a valid signature to the certificate of which it is the purported issuer, and that the signature has not yet expired.  This is not a function in which is reasonable or possible to attempt to second guess trust decisions made during chain construction.

**Comment by vdukhovni** [comment]
I am thrilled to see additional attention, review and suggested improvements to the X509 verification code. This is healthy for the project.  I applaud @DDvO's deep dive into the code.  However, I'd like to suggest less haste and more care, the reasons why things the way they are may not always be obvious, and we could inject even more prose (detailed rationale comments) into the code, but the non-obvious is not always a bug, and hasty changes not always progress.

**Comment by DDvO** [comment]
> I am thrilled to see additional attention, review and suggested improvements to the X509 verification code. This is healthy for the project. I applaud @DDvO's deep dive into the code. 

Pleased to hear!

> However, I'd like to suggest less haste and more care,

With the new issues and PRs I did not mean to imply rush.

> the reasons why things the way they are may not always be obvious, and we could inject even more prose (detailed rationale comments) into the code, 

Yes, also the mostly internal functions need more documentation, be it at code level and/or elsewhere. 
Even better would be if the names (of functions and important variables) were named more to-the-point and the code was better structured, such that the implementation becomes more readable.

> but the non-obvious is not always a bug, and hasty changes not always progress.

True. 
I'm perfectly fine if, and also have the impression myself that, more time is needed to sort things out.

**Comment by DDvO** [comment]
> > * While working myself through the code I recently found just one potential issue here: I think it would be good if chain verification re-checked the trust placed in the root of the chain, such that the security of chain verification becomes is independent of the checks done during chain building. This is requested to be added in #13770.
> 
> This is a mistake, resulting from a misunderstanding of the role of the `internal_verify()` function.
> Its purpose is solely to check chain _integrity_. That is, to ensure that the chain is actually a chain in which each issuer has applied a valid signature to the certificate of which it is the purported issuer, and that the signature has not yet expired. This is not a function in which is reasonable or possible to attempt to second guess trust decisions made during chain construction.
 
Then the "mistake" here is simply that I misunderstood the scope of `internal_verify()`.

Yet this does not invalidate my original point:
It must be possible, and IMHO should be done, to verify _independently of chain building_ that the root of a chain is trusted!

The chain building algorithm is necessarily much more complex than the chain verification, so it is more likely to include mistakes, also regarding the selection of the trust anchor.
Moreover, suppose that the chain building was provided by some external oracle, then it would be even more advisable to verify that the root of the proposed chain is correctly trusted.

If `internal_verify()` is not the right/optimal place to do it, where else is it then? 
Maybe at the end of `verify_chain()` (which BTW has a misleading name because it not only verifies the chain but also includes the chain building) ?
Or is the root trust (double-)check already done at some other point after `build_chain()` claimed success and I just failed to find it?

**Comment by vdukhovni** [comment]
> It must be possible, and IMHO should be done, to verify _independently of chain building_ that the root of a chain is trusted!

With bare public keys [1], there need not be any trusted certificates in the constructed chain.  And with PKIX-TA(0), there may need to be two-separate sources of trust to validate the chain.  And we only build as much of the chain as is required to acquire all the requisite trust-anchors, so I'm afraid that build_chain() is it, and if it is fragile or unclear, let's fix that.

The 1.1.0+ chain construction code is a substantial improvement over the code in 1.0.2.  Perhaps not enough, but it was a conservative effort to not make radical semantic departures, and so it supports fallback from trusted-first, partial-chain and other legacy variants. The only major change is that trusted-first is on by default, and no longer optional.

It is not accurate to say that DANE is a significant complication in the code, it is integrated in a fairly light-weight way, with DANE only really influencing the code in check_trust(), and disabling trust-store lookups when not doing PKIX validation (DANE-TA(2)).  Even with DANE, the new code is substantially less fragile than the old.  And, as you saw, the DANE code paths are reasonably well exercised by the test-suite (as are most other issues).

Where more help is needed is in the proxy-cert support, and CRL support.  I never had the cycles to deep-dive into those use-cases, and so they are essentially unchanged from 1.0.x and likely even longer ago.  It would be great if someone paid some attention to these, more than build_chain() which is in reasonably good shape.

For lack of a real intermediate CA store, and sufficiently clear/deep docs, users may have expectations of the semantics of CAfile/CApath that don't quite match the intent/reality.  That's unfortunate I guess, and better documentation should be the first step, and then some serious discussion of strategy: do we double-down on the status-quo and just dig the hole (lack of explicit trust) deeper, or is it time to instead offer a better cert store, perhaps after looking at what nss and GnuTLS do, and what users use/don't use or like/don't like about those.  Other comparisons may be with the Java trust store, MacOS System cert store, ...  There are many good precedents, and OpenSSL is fairly primitive in this space, not much has changed since 1999 when all the certs on the Web were signed directly by one or two root CAs (Verisign and later also Thawte).

[1] Presently via DANE-TA(2), but these need not originate in DNS, and can be supplied by the application, and indeed Postfix supports alternatives to WebPKI via this approach.

**Comment by DDvO** [comment]
@vdukhovni recently gave a comment in a related PR: https://github.com/openssl/openssl/pull/14199#discussion_r659122358
that IMO deserves high (though not urgent) attention and is better further discussed in the contents of this issue, so I quote it here:

>The certificate verification code is somewhat irreducibly complex, and one needs to be extremely cautious and detail oriented when working on it. Read, re-read, take notes, look up and down the call tree, ... and if something is unclear, don't change it.

I fully agree that one needs to be extremely cautious with the certificate patch building and validation topic.
Yet I only partially agree that it is somewhat irreducibly complex.

Sure the topic is non-trivial, but an important reason why one needs to be very careful is simply because, as already stated in the OP above, that its OpenSSL implementation is still needlessly messy and its documentation there is still needlessly poor.

**Comment by DDvO** [comment]
Over the last about 2 years a number of PRs have been brought up that aimed at improving the situation, with so far only limited success because those involved in them (i.e, the author(s) and the reviewers) were not able to spend sufficient effort on this.

BTW, in an analysis I did in aftermath of CVE-2021-3450 I found some fishy things that I have not even found the time to report yet.

I fear that the whole implementation needs a major overhaul and at least should be thoroughly re-reviewed, better structured, and systematically documented.

The only thing we have a chance to do for the upcoming 3.0 release is to further improve the documentation of the status quo,
and what I aim and propose to do is to fix by then the remaining mistiness in the user-oriented docs such as `openssl-verification-options.pod`.

**Comment by nhorman** [comment]
@DDvO This strikes me as an investiagtion issue, as we discussed previously, do you agree?
